### ISSUE TITLE
`GetOptionalId` is not consistent between implementations

### ISSUE THREAD

**Issue by HarelM** [NetTopologySuite/NetTopologySuite.Features#17]
When creating a "regular" feature, serializing it and then deserializing it, I get a different behavior for `GetOptionalId`.
Here is the relevant test:
```cs
        [TestMethod]
        public void CheckGetOptionalIdEquality()
        {
            var options = new JsonSerializerOptions();
            options.Converters.Add(new GeoJsonConverterFactory());

            var feature = new Feature(new Point(0, 0), new AttributesTable { {"some-id", "my-id"} });
            var fullCycleFeature = JsonSerializer.Deserialize<IFeature>(JsonSerializer.Serialize(feature, options), options);
            Assert.AreEqual(feature.GetOptionalId("some-id"), fullCycleFeature.GetOptionalId("some-id"));
        }
 ```

This bit me hard in one of the releases I made :-(
I think it has to do with the IUniquId part of the STJ feature implementation.

**Comment by FObermaier** [comment]
As I'm revisiting and -investigating this issue I begin to think that the problem is that the `StjFeatureConverter` elevates "some-id" to `IUnique.Id` and the `StjAttributeTableConverter` omits "some-id" when writing properties.

This would make it an issue with NetTopologySuite.IO.GeoJSON. @airbreather what do you think?

I'd propose to only set `IUnique.Id` when the property name "id" is present in json and leave the properties as they are.

**Comment by airbreather** [comment]
> As I'm revisiting and -investigating this issue I begin to think that the problem is that the `StjFeatureConverter` elevates "some-id" to `IUnique.Id` and the `StjAttributeTableConverter` omits "some-id" when writing properties.
> [...]
> I'd propose to only set `IUnique.Id` when the property name "id" is present in json and leave the properties as they are.

I disagree. Instead, I strongly believe that this is "working as intended". I have a wall of text that I'll condense below (I really don't want to delete all of it, especially the "History" section), because the takeaway is actually very simple.

If you don't want to (or can't) match the `idPropertyName` used for `GetOptionalId` and `GeoJsonConverterFactory`, then I see no reason why you should be using `GetOptionalId` in the first place.

What problem do you encounter when using `IFeature.Attributes` directly? Depending on what the problem is, it may also help to understand why you expect `GetOptionalId` to fix that problem?

<details>
<summary>Wall of text: why I think this is "working as intended"</summary>

Most obviously, the documentation clearly states that this is how it works, so if I were a more dismissive person, I'd just leave it here: https://github.com/NetTopologySuite/NetTopologySuite.Features/blob/0cb0a03dba8028fde185601168eeb652c3699a50/src/NetTopologySuite.Features/FeatureExtensions.cs#L8-L12

I'll try to be more helpful, though, because I certainly wouldn't be satisfied by that.

Serialization and deserialization are two different and somewhat independent things, sometimes **very** different and **very** independent when both the format (RFC 7946 GeoJSON) **and** the object model (NetTopologySuite.Features, used for more than just GeoJSON) are constrained externally. Not everything that should be modeled by `IFeature` has an "ID" as such, so `IFeature` doesn't have anything of the sort.

Of course, a GeoJSON `Feature` **does** have an optional `"id"`, so the goal was to make it so that when you deserialize a GeoJSON `Feature`, you receive an object where you could tell whether or not it had an `"id"` (with the one limitation being that you may not necessarily be able to tell the difference between "present, but set to `null`" and "not present at all"). So that's the deserialization side accounted for.

On the serialization side, the goal was to make it so that you can use the straightforward `Feature` class for writing out a GeoJSON `Feature` object. For that, the idea was to have you put a special attribute into its `Attributes` that the JSON converters would interpret as "this one is special: don't put it into the `"properties"`, but make it the `"id"` instead. This lets you write out any RFC 7946 GeoJSON object that you want, unless you **REALLY** want to write something in there with the "special" property name... and even then, you can override which property name is "special". All right, cool, serialization is sorted now.

So let's go through the code sample at the top with that in mind. I'll insert some in-between variables for the inline expressions.

```csharp
var options = new JsonSerializerOptions();
options.Converters.Add(new GeoJsonConverterFactory());
```

Makes sense so far: you don't want anything special, just the default out-of-the-box behavior.

```csharp
var feature = new Feature(new Point(0, 0), new AttributesTable { {"some-id", "my-id"} });
string serialized = JsonSerializer.Serialize(feature, options);
```

Nothing wrong here. Looks to me like this should serialize as something equivalent to:

```json
{
  "type": "Feature"
  "properties": {
    "some-id": "my-id"
  },
  "geometry": {
    "type": "Point",
    "coordinates": [0, 0]
  }
}
```

It doesn't have an `"id"`, and I don't think anyone expected it to have an `"id"` yet. All right, let's move on.

```csharp
var fullCycleFeature = JsonSerializer.Deserialize<IFeature>(serialized, options);
```

No drama here, either: you should get an `IFeature` whose `Geometry` is that `Point` and whose `Attributes` contains a single attribute, named `"some-id"`, with the value `"my-id"`.

```csharp
object idBefore = feature.GetOptionalId("some-id");
```

Here's where I start to see a problem. Why are you using `GetOptionalId` for this? In this sample, you have an object with a `"some-id"` attribute. You can just use `feature.Attributes["some-id"]`. Furthermore, the variable is even a `Feature` type (not `IFeature`), so there's no way that it can be different.

This is highly suspect, and without any further context for what you are trying to achieve, it seems like there's a much more straightforward way to fix this code sample than opening an issue on this issue tracker and waiting for a resolution: use `feature.Attributes`. That's what you used to set it in the first place, after all.

```csharp
object idAfter = fullCycleFeature.GetOptionalId("some-id");
```

This is where it goes from "highly suspect" to "wrong unless proven otherwise". You have an `IFeature` that came from what you **know** to be GeoJSON in this context. No matter what context I imagine that this code is running in, I must assume that it **at least reasonably suspects** that the `IFeature` **might** have come from GeoJSON.

So although you don't control the code in either the `NetTopologySuite.Features` or `NetTopologySuite.IO.GeoJSON` GitHub projects, I do expect that you control the code that produces the `GeoJsonConverterFactory` and the code that calls `GetOptionalId`. Here, `GetOptionalId` (specifically, `GetOptionalId(GeoJsonConverterFactory.DefaultIdPropertyName)` because you didn't override it when constructing the `GeoJsonConverterFactory`) is there to let you get the `"id"` that was originally there in the GeoJSON. I have yet to see any arguments whatsoever explaining why it should do something different here. The normal way to access a value from an `IFeature.Attributes` is to use `IFeature.Attributes`. `GetOptionalId` has different semantics.

</details>

<details>
<summary>History: how we got here in the first place</summary>

The intention was that in the overwhelming majority of cases, only the `"_NetTopologySuite_id"` property (also called `GeoJsonConverterFactory.DefaultIdPropertyName`) would get this special treatment: it's extremely unlikely that someone would actually have a use for a property of this name, so it lets us set the GeoJSON `Feature`'s top-level `"id"` property using a regular old `Feature` object.

The `Newtonsoft.Json` converters work similarly, with the main differences being:
- instead of the property `"_NetTopologySuite_id"` getting special treatment, it's merely `"id"`. there is no way to override which property gets this special treatment, it must be `"id"`
- technically, there's a global `static` variable called `AttributesTableConverter.WriteIdToProperties` which you can override to `true` for the entire application so that an `"id"` will still get written to the GeoJSON `Feature`'s `"properties"` object in addition to its top-level `"id"` property.

What this means in that universe is that if you have a GeoJSON document that uses the `Feature`'s `"id"` and its `"properties"` object's `"id"` for two different things, then it's simply impossible to model this same document with the `Newtonsoft.Json` converters:

```json
{
  "type": "Feature",
  "id": 123,
  "properties": {
    "id": 456,
    "name": "some name"
  },
  "geometry": {
    "type": "Point",
    "coordinates": []
  }
}
```

I can **easily** imagine this happening with a system where the `"properties"` is a serialized representation of some arbitrary domain object (which can easily have its own ID property with its own semantics), and the top-level `"id"` is supposed to be used for something totally different.

IIRC, when preparing NetTopologySuite/NetTopologySuite.IO.GeoJSON#50, I almost didn't want to make it possible to override `"_NetTopologySuite_id"` because it seemed way too unlikely that someone would actually need to use a different property for this purpose. However, I believe that I envisioned one main use case for overriding it:
- Someone moving from the `Newtonsoft.Json`-based converters may already have been using `"id"` for this throughout their project.
- Letting those projects opt-in like this would make the transition slightly easier without forcing this surprising `"id"` property behavior onto everybody else.

</details>

**Comment by HarelM** [comment]
I've opened this issue a long time ago, I'll see if I can recall the idea behind what I tired to do.
I obviously solved the issue by using the `Attributes`, I didn't wait for a fix but rather wanted to inform you that there is something that is not intuitive with the current API/implementation.
Here's my setup:
I have a GeoJson stored in a elasticsearch database, in which I have no control over its Factory, currently using STJ.
I have old features in the database that doesn't have `id` on the feature, but have `poiId` property in the attributes.
I have some other features which are stored in a CSV file, and gets converted to `Feature` without being geojson at any point.
I wanted to be able to transition the database from using the `poiId` to using `id` gradually - without needing to reindex or anything else, just new features that I write will have the `id` and keep the old ones with `poiId`, until there are no more in the database.
In my code (not in the data access layer) all I know is that I have `IFeature`, I don't know where it came from and if it uses old id (`poiId`) or new ID (`id` on the feature).

My intuition with a method that is called `getOptinalId` is to get the id from the feature if that exists, or get it from the `Attributes` if it doesn't (according to the relevant attribute - `poiId` in my case).
But my intuition was wrong, which led me to open this issue.

You are obviously free to ignore it, and you can also say: hey, you can simply create an extension method which does exactly that instead of using the exiting API, and that's fine too.

The bottom line is that I can't use `getOptinalId` due to implementation details. Yes the documentation is stating something about internal implementation of `IUniqueId` which I have no good understanding of why it is used for `IFeature` and not for `StjFeature` or the other way around...

I hope this gives a bit more context to the initial issue.

Please don't read the above as angry or aggressive in any way, I truly admire the amazing work you are doing and I just wanted to bring some minor DX to your attention.

**Comment by FObermaier** [comment]
OK, we'll leave it as is.
